xdsl/ir/core.py

```diff
@@ -708,6 +708,33 @@ def __hash__(self):
         return hash(self._op._operands)  # pyright: ignore[reportPrivateUsage]
 
 
+class OpTraits(Iterable[OpTrait]):
+    _traits: set[OpTrait] | Callable[[], set[OpTrait]]
+
+    def __init__(self, traits: set[OpTrait] | Callable[[], set[OpTrait]]) -> None:
+        self._traits = traits
+
+    def get_traits(self) -> set[OpTrait]:
+        """Returns a copy of this instance's traits."""
+        if not isinstance(self._traits, set):
+            self._traits = self._traits()
+        return set(self._traits)
+
+    def add_trait(self, trait: OpTrait):
```

Comment:
Is this dead code?

Reply:
Tested here, but may well be dead otherwise...
```
def test_modify_traits():
    op = TestModifyTraitsOp()

    op.verify()

    TestModifyTraitsOp.traits.add_trait(AlwaysFailsTrait())

    with pytest.raises(VerifyException, match="Nope"):
        op.verify()
```

Reply:
I would vote to get rid of it, as I see no reason for operation definitions to be mutable, but will leave it up to you.

Reply:
I would actually like to keep it, and I have some use cases for it.
If you are a downstream user, and you want to add an interface to an upstream dialect (let say, adding a semantic interface to an existing operation), then you need to actually modify the trait interfaces.